src/data/finsupp/lex.lean

```diff
@@ -0,0 +1,145 @@
+  | (wit : α), h :=
+    have hne : f wit ≠ g wit := mem_ne_locus.mp (finset.mem_of_min h),
+    hne.lt_or_lt.by_cases
+      (λ hwit, h_lt ⟨wit, λ j hj, mem_ne_locus.not_left.mp (finset.not_mem_of_lt_min hj h), hwit⟩)
+      (λ hwit, h_gt ⟨wit, by exact λ j hj, begin
+        refine mem_ne_locus.not_left.mp (finset.not_mem_of_lt_min hj _),
+        rwa ne_locus_comm,
+      end, hwit⟩)
+  end
+
+instance lex.decidable_le : @decidable_rel (lex (α →₀ N)) (≤) :=
+lt_trichotomy_rec
+  (λ f g h, is_true $ or.inr h)
+  (λ f g h, is_true $ or.inl $ congr_arg _ h)
+  (λ f g h, is_false $ λ h', (lt_irrefl _ (h.trans_le h')).elim)
+
+instance lex.decidable_lt : @decidable_rel (lex (α →₀ N)) (<) :=
+lt_trichotomy_rec
+  (λ f g h, is_true h)
+  (λ f g h, is_false h.not_lt)
+  (λ f g h, is_false h.asymm)
+
+/--  The linear order on `finsupp`s obtained by the lexicographic ordering. -/
+instance lex.linear_order : linear_order (lex (α →₀ N)) :=
+{ le_total := lt_trichotomy_rec
+    (λ f g h, or.inl h.le)
+    (λ f g h, or.inl h.le)
+    (λ f g h, or.inr h.le),
+  decidable_lt := by apply_instance,
+  decidable_le := by apply_instance,
+  decidable_eq := by apply_instance,
+  ..lex.partial_order }
+
+lemma lex.le_of_forall_le {a b : lex (α →₀ N)} (h : ∀ i, of_lex a i ≤ of_lex b i) : a ≤ b :=
+le_of_not_lt (λ ⟨i, hi⟩, (h i).not_lt hi.2)
+
+lemma lex.le_of_of_lex_le {a b : lex (α →₀ N)} (h : of_lex a ≤ of_lex b) : a ≤ b :=
+lex.le_of_forall_le h
+
+lemma to_lex_monotone : monotone (@to_lex (α →₀ N)) :=
+λ _ _, lex.le_of_forall_le
+
+end N_has_zero
+
+section inhabited_and_linear_orders
+variables [inhabited α] [linear_order α] [linear_order N] [has_zero N] {f g : α →₀ N}
+
+/--  This is a technical result.  Likely, you will need one of the consequences of this lemma.  -/
+lemma apply_wit_lt_apply_wit_of_to_lex_lt (fg : to_lex f < to_lex g) :
+  f (f.wit g) < g (f.wit g) :=
+begin
+  rcases fg with ⟨x, ltx, fgx⟩,
+  { convert fgx;
+    exact wit_eq_of_ne_of_forall fgx.ne ltx }
+end
+
+lemma to_lex_le_iff_apply_wit_le : to_lex f ≤ to_lex g ↔ f ((f.wit g)) ≤ g ((f.wit g)) :=
+begin
+  refine ⟨λ h, _, λ h, _⟩,
+  { rcases eq_or_ne f g with rfl | fg,
+    { exact le_rfl },
+    { exact (apply_wit_lt_apply_wit_of_to_lex_lt (h.lt_of_ne fg)).le } },
+  { contrapose! h,
+    rw wit_comm,
+    exact apply_wit_lt_apply_wit_of_to_lex_lt h }
+end
+
+lemma to_lex_lt_iff_apply_wit_lt : to_lex f < to_lex g ↔ f (f.wit g) < g (f.wit g) :=
+not_iff_not.mp (by simp only [to_lex_le_iff_apply_wit_le, wit_comm, not_lt])
+
+lemma le_iff_of_lex_apply_wit_le {f g : lex (α →₀ N)} :
+  f ≤ g ↔ of_lex f (f.wit g) ≤ of_lex g (f.wit g) :=
+by simp [of_lex, to_lex, ← to_lex_le_iff_apply_wit_le]
+
+end inhabited_and_linear_orders
+
+variables [linear_order α] [linear_order N] [add_left_cancel_monoid N] [covariant_class N N (+) (≤)]
+
+instance : covariant_class (lex (α →₀ N)) (lex (α →₀ N)) (@has_add.add (α →₀ N) _) (≤) :=
```

Comment:
This statement looks pretty suspect to me. If we want to go down this route, we should copy all algebraic structure onto `lex` rather than hack around it with `@`. Thoughts, @YaelDillies?

Reply:
We sure want the group structure on `lex`, because it will be an ordered group! This is already done for `pi.lex`.

Reply:
We should definitely generalize those then such that `lex foo` has the same algebra structure (or at least, additive structure for now) as `foo`

Reply:
For my intended application, `α →₀ N` is going to be the grading additive monoid of an `add_monoid_algebra`, for instance, `mv_polynomial`.  In this context, `N` will have all the order and covariant_class assumptions, whereas `α` is going to be a "bare" Type.

What I will want to do is to endow `α` with *any* linear order, induce the order and the covariant assumptions on `α →₀ N` and use them to prove whatever result I need (no zero divisors in the immediate application).

So, to summarize, just the additive structure is enough for my purposes.  I currently do not think that I have a use-case for wanting, e.g. a semiring structure on `lex finsupp`.

Reply:
Are we talking of adding what is below and more?
```
noncomputable instance [linear_order α] [add_monoid N] [linear_order N] :
  add_monoid (lex (α →₀ N)) :=
finsupp.add_monoid

noncomputable instance [linear_order α] [add_comm_monoid N] [linear_order N] :
  add_comm_monoid (lex (α →₀ N)) :=
finsupp.add_comm_monoid

noncomputable instance [linear_order α] [add_group N] [linear_order N] :
  add_group (lex (α →₀ N)) :=
finsupp.add_group

noncomputable instance [linear_order α] [add_comm_group N] [linear_order N] :
  add_comm_group (lex (α →₀ N)) :=
finsupp.add_comm_group

...
```

Reply:
No, I mean adding
```
noncomputable instance [add_monoid N] :
  add_monoid (lex N) :=
‹add_monoid N›
```

Reply:
Oh, I see!  Adding the instances way before the linear order!

Reply:
That should be computable, right? 😅

Reply:
So, something like this?
```
@[to_additive]
instance [has_one N] : has_one (lex N) := ‹has_one N›
@[to_additive]
instance [mul_one_class N] : mul_one_class (lex N) := ‹mul_one_class N›
@[to_additive]
instance [semigroup N] : semigroup (lex N) := ‹semigroup N›
@[to_additive]
instance [monoid N] : monoid (lex N) := ‹monoid N›
@[to_additive]
instance [comm_monoid N] : comm_monoid (lex N) := ‹comm_monoid N›
@[to_additive]
instance [group N] : group (lex N) := ‹group N›
@[to_additive]
instance [left_cancel_monoid N] : left_cancel_monoid (lex N) := ‹left_cancel_monoid N›
@[to_additive]
instance [right_cancel_monoid N] : right_cancel_monoid (lex N) := ‹right_cancel_monoid N›
@[to_additive]
instance [cancel_monoid N] : cancel_monoid (lex N) := ‹cancel_monoid N›
```

Reply:
Is there a way of telling Lean to do the above for all classes that do not contain a field called `le` or `lt`?

Reply:
Not really, no...

Reply:
Don't forget to write the lemmas of the form `to_lex 0 = 0`, `of_lex 0 = 0`, etc...

Reply:
This is probably a good point to spin off a new PR

Reply:
Eric, surely I would not have forgotten such lemmas...  :roll_eyes:

Reply:
> This is probably a good point to spin off a new PR

Yes, I am also still mulling over ditching `wit`, so I will extract a smaller portion of this one to one that does not depend on witnesses, but only on `ne_loci`.

Reply:
Would it make sense to create a folder `order.lex` and place there all the `lex` stuff?

Reply:
Thus, have files `order.lex.lex, order.lex.algebra, order.lex.pi, order.lex.finsupp`...?

Reply:
No, probably not. Let me open the PR adding those typeclasses quick.

Reply:
Sure!  If it helps, this is what I have:

```lean
variables {α : Type*}
namespace lex

@[to_additive]
instance [has_one α] : has_one (lex α) := ‹has_one α›

@[to_additive] lemma to_lex_one [has_one α] : to_lex 1 = 1 := rfl
@[to_additive] lemma of_lex_one [has_one α] : of_lex 1 = 1 := rfl

@[to_additive]
instance [has_mul α] : has_mul (lex α) := ‹has_mul α›
@[to_additive]
instance [mul_one_class α] : mul_one_class (lex α) := ‹mul_one_class α›
@[to_additive]
instance [semigroup α] : semigroup (lex α) := ‹semigroup α›
@[to_additive]
instance [comm_semigroup α] : comm_semigroup (lex α) := ‹comm_semigroup α›
@[to_additive]
instance [monoid α] : monoid (lex α) := ‹monoid α›
@[to_additive]
instance [comm_monoid α] : comm_monoid (lex α) := ‹comm_monoid α›
@[to_additive]
instance [group α] : group (lex α) := ‹group α›
@[to_additive]
instance [comm_group α] : comm_group (lex α) := ‹comm_group α›
@[to_additive]
instance [left_cancel_semigroup α] : left_cancel_semigroup (lex α) := ‹left_cancel_semigroup α›
@[to_additive]
instance [left_cancel_monoid α] : left_cancel_monoid (lex α) := ‹left_cancel_monoid α›
@[to_additive]
instance [right_cancel_monoid α] : right_cancel_monoid (lex α) := ‹right_cancel_monoid α›
@[to_additive]
instance [right_cancel_semigroup α] : right_cancel_semigroup (lex α) := ‹right_cancel_semigroup α›
@[to_additive]
instance [cancel_monoid α] : cancel_monoid (lex α) := ‹cancel_monoid α›
@[to_additive]
instance [cancel_comm_monoid α] : cancel_comm_monoid (lex α) := ‹cancel_comm_monoid α›
@[to_additive]
instance [has_involutive_inv α] : has_involutive_inv (lex α) := ‹has_involutive_inv α›
@[to_additive]
instance [div_inv_monoid α] : div_inv_monoid (lex α) := ‹div_inv_monoid α›
@[to_additive]
instance [division_monoid α] : division_monoid (lex α) := ‹division_monoid α›
@[to_additive]
instance [division_comm_monoid α] : division_comm_monoid (lex α) := ‹division_comm_monoid α›

end lex
```

Reply:
#16122